workflows/test_validate.py

```diff
@@ -0,0 +1,130 @@
+"""
+Tests the functionality of validate.py with several different recipes which should raise different errors
+"""
+
+import pytest
+import mock
+import workflows
+from workflows.validate import validate_recipe, main
+import json
+import sys
+
+
+def test_no_file():
+    """Test that validate returns an error when it returns with no file"""
+    with pytest.raises(Exception):
+        validate_recipe()
+
+
+def test_healthy_recipe():
+    """Test that no errors are raised when given a healthy recipe"""
+    healthy_recipe = """
+        {
+          "1": {
+            "queue": "simpleservice.submission",
+            "parameters": {
+              "commands": [
+                "echo This is a command32"
+              ],
+              "workingdir": "/dls/tmp/riw56156/zocalo",
+              "output_file": "out.txt"
+            }
+          },
+          "start": [
+            [1, []]
+          ]
+        }
+    """
+
+    # Run validate with mock open, expect no exceptions
+    with mock.patch(
+        "builtins.open", mock.mock_open(read_data=healthy_recipe), create=True
+    ) as mock_file:
+        validate_recipe("/path/to/open")
```

Comment:
Instead of having to patch `open` (which is possible) you can just make up a normal file.
Add parameter `tmpdir` to your test, as in 
```python
def test_...(tmpdir):
``` 
then write the actual file
```python
recipe_file = tmpdir.join("my_file.json")
recipe_file.write(healthy_recipe)
```
and finally:
```python
validate_recipe(recipe_file.strpath)
```

Reply:
Doesn't this go against the general unit test principles? And would involve writing a file to disk on each unit test that requires it? Then deleting it?

Reply:
fyi if you haven't come across that before: [`tmpdir` is a pytest fixture](https://docs.pytest.org/en/latest/tmpdir.html)
`tmpdir` is a `py.path` object and offers [a bunch of useful functions](https://py.readthedocs.io/en/latest/path.html).

I also just saw that there is now also `tmp_path` which uses the Python pathlib standard library. I would stick with `tmpdir` for now due to our legacy commitments. :roll_eyes:

Reply:
Depends on how unit-y you want your test to be. Your function reads a file, so I don't see why you shouldn't be able to give it a file.

There is no need to delete the file at the end of the test. pytest puts it into a temporary directory that is managed by pytest, so it will disappear eventually.

Reply:
Ok no, didn't know about that, will follow up and add this in